Fixes #38219 - Update Ansible Tower API URL for AAP 2.5+ compatibility

nofaralfasi opened the pull request:
AAP 2.5+ uses `/api/controller/v2` instead of `/api/v2`.
This updates provisioning templates to use `ansible_tower_api_url`
and includes a migration renaming `ansible_tower_fqdn`, appending the new API path.

Docs PR (upgrade warning): https://github.com/theforeman/foreman-documentation/pull/3720

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

Gauravtalreja1 commented on app/views/unattended/provisioning_templates/snippet/ansible_tower_callback_service.erb:
```diff
@@ -7,6 +7,15 @@ description: |
+
+<%
+  api_base = if host_param('ansible_tower_version').to_f >= 2.5
+               '/api/controller/v2'
```
```suggestion
               '/api/controller/v2/job_templates'
```

nofaralfasi replied:
Thank you for noticing!

nofaralfasi commented on app/views/unattended/provisioning_templates/snippet/ansible_tower_callback_script.erb:
```diff
@@ -10,9 +10,17 @@ description: |
 echo "Calling Ansible AWX/Tower provisioning callback..."
+
+# Query the AC version
+if [ $(echo "$(curl -k -s https://<%= host_param('ansible_tower_fqdn') %>/api/v2/ping/ | jq -r .version | cut -d. -f1,2) >= 4.6" | bc) -eq 1 ]; then
```
Do I need to use `/usr/bin/curl` here, as in the lines below?

sayan3296 replied:
It makes sense to follow the same pattern i.e. use of the full binary path. So +1 ..

evgeni commented on app/views/unattended/provisioning_templates/snippet/ansible_tower_callback_script.erb:
```diff
@@ -10,9 +10,17 @@ description: |
 echo "Calling Ansible AWX/Tower provisioning callback..."
+
+# Query the AC version
+if [ $(echo "$(curl -k -s https://<%= host_param('ansible_tower_fqdn') %>/api/v2/ping/ | jq -r .version | cut -d. -f1,2) >= 4.6" | bc) -eq 1 ]; then
```
This runs on the machine being deployed, so I don't think we can rely on `jq` being available

sayan3296 replied:
Maybe something like this to replace the `jq -r .version` part? 
```
/usr/bin/python3 -c 'import json, sys; input_data = sys.stdin.read().strip(); print(json.loads(input_data).get("version", 0) if input_data else 0)'
```

Feels a bit odd to suggest putting a long one-liner in a snippet

ekohl replied:
Probably very fragile, but would it make sense to query the version on the Foreman side and then render the correct template?

Another thought: if you're going to rely on `/usr/bin/python3` anyway, rewrite the whole script to Python instead of bash?

Third option: first post to the new API endpoint and if that fails, post to the old API endpoint.

nofaralfasi replied:
I think I prefer the third option.

evgeni replied:
The new implementation (while untested) fixes my concerns, thanks!

Gauravtalreja1 commented on app/views/unattended/provisioning_templates/snippet/ansible_tower_callback_script.erb:
```diff
@@ -10,9 +10,17 @@ description: |
 echo "Calling Ansible AWX/Tower provisioning callback..."
+
+# Query the AC version
+if [ $(echo "$(curl -k -s https://<%= host_param('ansible_tower_fqdn') %>/api/v2/ping/ | jq -r .version | cut -d. -f1,2) >= 4.6" | bc) -eq 1 ]; then
```
Just curious to know if `/api/v2/ping/` is same endpoint for both versions?

sayan3296 replied:
I have a 4.5.8 where I see the ping endpoint works.. Not sure about even older instances though.

nofaralfasi commented:
The failing tests seem related. I'm actively working on fixing them, but I'd appreciate a review of the overall implementation.

nofaralfasi commented:
Tests are fixed. The remaining failures are unrelated to this PR.

nofaralfasi commented:
@evgeni can you please approve this PR so we'll have at least one approval?

evgeni commented:
The changes look sane to me, but I have no environment to test this, sorry.

Gauravtalreja1 commented on app/views/unattended/provisioning_templates/snippet/ansible_tower_callback_service.erb:
```diff
@@ -15,9 +26,9 @@ After=network-online.target
 Type=oneshot
 <% if host_param('ansible_extra_vars') -%>
-ExecStart=/usr/bin/curl -k -s -H 'Content-Type: application/json' --data '{"host_config_key":"<%= host_param('ansible_host_config_key') %>", "extra_vars": <%=host_param('ansible_extra_vars') %>}' https://<%= host_param('ansible_tower_fqdn') %>/api/v2/job_templates/<%= host_param('ansible_job_template_id') %>/callback/
+ExecStart=/usr/bin/curl -k -s -H 'Content-Type: application/json' --data '{"host_config_key":"<%= host_param('ansible_host_config_key') %>", "extra_vars": <%=host_param('ansible_extra_vars') %>}' https://<%= host_param('ansible_tower_fqdn') %>${api_base}<%= host_param('ansible_job_template_id') %>/callback/
```
I tested this change and here `${api_base}` doesn't render at all, so not sure if it should be `<%= api_base %> or something else in this erb template

Gauravtalreja1 commented on app/views/unattended/provisioning_templates/snippet/ansible_tower_callback_service.erb:
```diff
@@ -7,6 +7,17 @@ description: |
 -%>
+
+# Attempt to use the /api/controller/v2 API for AAP 2.5 and higher. If successful, continue with newer API endpoints.
+response=$(/usr/bin/curl -k -s --write-out "%{http_code}" -o /dev/null https://<%= host_param('ansible_tower_fqdn') %>/api/controller/v2/)
```
This adds following warnings in systemd logs for `ansible-callback.service`
```
Mar 05 08:28:49 toby-steever.example.com systemd[1]: /etc/systemd/system/ansible-callback.service:3: Assignment outside of section. Ignoring.
Mar 05 08:28:49 toby-steever.example.com systemd[1]: /etc/systemd/system/ansible-callback.service:6: Assignment outside of section. Ignoring.
Mar 05 08:28:49 toby-steever.example.com systemd[1]: /etc/systemd/system/ansible-callback.service:7: Assignment outside of section. Ignoring.
Mar 05 08:28:49 toby-steever.example.com systemd[1]: /etc/systemd/system/ansible-callback.service:8: Assignment outside of section. Ignoring.
Mar 05 08:28:49 toby-steever.example.com systemd[1]: /etc/systemd/system/ansible-callback.service:9: Assignment outside of section. Ignoring.
Mar 05 08:28:49 toby-steever.example.com systemd[1]: /etc/systemd/system/ansible-callback.service:10: Assignment outside of section. Ignoring.
.
.
```

evgeni replied:
oh yes, I only looked at the `callback_script.erb` changes, which is a shell script.

the same doesn't work in a systemd service :see_no_evil:

ekohl commented:
From Matrix: you can also introduce a parameter to determine the version. @evgeni then suggested to take it a step further and instead of using `ansible_tower_fqdn` as the parameter to use `ansible_tower_api_url`. I like that because any network request can (and will) fail. Your autodetection needs to handle that and making that reliable is hard.

nofaralfasi commented:
> @evgeni then suggested to take it a step further and instead of using `ansible_tower_fqdn` as the parameter to use `ansible_tower_api_url`. I like that because any network request can (and will) fail. Your autodetection needs to handle that and making that reliable is hard.

I suggest introducing a new parameter, `ansible_tower_api_url`, to store the full API path while retaining `ansible_tower_fqdn` for backward compatibility. If `ansible_tower_api_url` is not provided, we default to using `ansible_tower_fqdn` with the new API path (`/api/controller/v2/`).  

I initially considered falling back to the old API (`/api/v2`), but I’m unsure if that’s the right approach. Any thoughts?

nofaralfasi committed:
Fixes #38219 - Update Ansible Tower API URL for AAP 2.5+ compatibility

AAP 2.5+ uses `/api/controller/v2` instead of `/api/v2`.
This updates provisioning templates to use `ansible_tower_api_url`
and includes a migration renaming `ansible_tower_fqdn`, appending the new API path.

evgeni commented:
I guess the alternative is to write a migration that does "if tower_fqdn is set, create tower_url with the right (new) path and delete fqdn". 
Only really works for global parameters, but probably good enough?

ekohl commented:
> I guess the alternative is to write a migration that does "if tower_fqdn is set, create tower_url with the right (new) path and delete fqdn".

Migration makes sense to me.

> Only really works for global parameters, but probably good enough?

Why wouldn't it work for host params?

ekohl commented on app/views/unattended/provisioning_templates/snippet/ansible_tower_callback_script.erb:
```diff
@@ -9,10 +9,12 @@ description: |
 -%>
 #!/bin/sh
 
+<% base_api = host_param('ansible_tower_api_url') || "#{host_param('ansible_tower_fqdn')}/api/controller/v2" %>
+
```
I'd suggest to keep the whitespace change to a minimum. You can see the difference in the generated snapshots.
```suggestion
<% base_api = host_param('ansible_tower_api_url') || "#{host_param('ansible_tower_fqdn')}/api/controller/v2" -%>
```

nofaralfasi commented:
> I initially considered falling back to the old API (/api/v2), but I’m unsure if that’s the right approach. Any thoughts?

One more thing I forgot to mention - the new API path `/api/controller/v2/` doesn’t fully work for AAP 2.4. Specifically, it works on the **Controller server** but **not on the Gateway server** in version 2.4.

nofaralfasi commented:
> > I guess the alternative is to write a migration that does "if tower_fqdn is set, create tower_url with the right (new) path and delete fqdn".
> 
> Migration makes sense to me.

Is changing host parameters via migrations something we usually do? We’d update the documentation accordingly, but I want to confirm if this is a valid approach before moving forward. Also, just to clarify, the new API path only doesn't work for AAP 2.4 and older versions.

evgeni commented:
> > Only really works for global parameters, but probably good enough?
> 
> Why wouldn't it work for host params?

My brain was going "these are hard to find" as they also can be defined on a domain, hostgroup, etc.
But the data is probably still in the same table? idk about params >.<

ekohl commented:
> Is changing host parameters via migrations something we usually do? We’d update the documentation accordingly, but I want to confirm if this is a valid approach before moving forward. Also, just to clarify, the new API path only doesn't work for AAP 2.4 and older versions.

There's an argument to be made for copying it, in case the user copied the template which doesn't have the changes.

Regardless, probably worth a release note.

nofaralfasi commented:
> There's an argument to be made for copying it, in case the user copied the template which doesn't have the changes.

I decided not to keep the old parameter name (`ansible_tower_fqdn`) since users are responsible for maintaining their own templates. Given that the new API path will eventually be required by all users, retaining the old parameter didn’t seem necessary.

ekohl commented on test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Kickstart_default_finish.host4dhcp.snap.txt:
```diff
@@ -227,7 +227,7 @@ cat << EOF-9f037ba1 > /root/ansible_provisioning_call.sh
 
 echo "Calling Ansible AWX/Tower provisioning callback..."
-/usr/bin/curl -v -k -s --data "host_config_key=" https:///api/v2/job_templates//callback/
+/usr/bin/curl -v -k -s --data "host_config_key=" https:///job_templates//callback/
```
Seeing this incomplete snapshot I wonder if we should make the template fail if the parameter is not set. Given we're renaming things, it may make sense.

Had we done this previously then any forked templates would also start to fail explicitly instead of silently because of the missing parameter.

Any thoughts on this?

nofaralfasi replied:
I agree, we should add that. However, I suggest doing it in a separate PR since this fix is more urgent, and we want to get it in as soon as possible.

ekohl replied:
Perhaps a start to improve test coverage is to at least set the parameter here:
https://github.com/theforeman/foreman/blob/b42ca08e990c99f32838997a90c805ac85fc7af2/app/services/foreman/template_snapshot_service.rb#L72

nofaralfasi replied:
Should we also add the `ansible_job_template_id` param here?

ekohl reviewed (dismissed):
> I decided not to keep the old parameter name (ansible_tower_fqdn) since users are responsible for maintaining their own templates. Given that the new API path will eventually be required by all users, retaining the old parameter didn’t seem necessary.

This is fine by me, but please submit a PR to add an upgrade warning to the release notes (https://github.com/theforeman/foreman-documentation/blob/master/guides/doc-Release_Notes/topics/foreman.adoc?plain=1). When you submit the PR, please set the label "Waiting for code". Though there's a chance you can't add labels but using a draft and clearly describing the context is likely sufficient.

Leaving PR open since there were other reviewers and to give you a chance to look at my (non-blocking) suggestion.

ekohl reviewed (commented):
I missed this before: the templates still contain `https://` as a prefix, but when I read a URL I also expect the scheme to be part of it. It's likely that users end up with `https://https://host.example.com`.

I'd suggest to drop `https://` from the templates and make it part of the parameter.

nofaralfasi commented on db/migrate/20250309121956_rename_ansible_tower_fqdn_to_api_url.rb:
```diff
@@ -0,0 +1,19 @@
+    Parameter.where(name: 'ansible_tower_api_url').find_each do |parameter|
+      parameter.update(
+        name: 'ansible_tower_fqdn',
+        value: parameter.value.sub('https://', '').sub('/api/controller/v2', '')
```
This approach wouldn't work for variables whose values have been modified. Should we also account for cases where the API path of the variable has been updated?

ekohl replied:
I'd use [URI](https://rubyapi.org/o/uri) to extract the host of the URL.
```suggestion
        value: URI(parameter.value).host
```

ekohl reviewed (approved):
I don't have the setup to verify this myself, but from reading the code & snapshots this looks correct to me.

Anyone else mind giving it a look?

shweta83 commented:
@nofaralfasi I have tested the new api endpoint and the migration with AAP2.5 and it worked.

shweta83 reviewed (approved):
ack

Gauravtalreja1 commented on db/migrate/20250309121956_rename_ansible_tower_fqdn_to_api_url.rb:
```diff
@@ -0,0 +1,19 @@
+    Parameter.where(name: 'ansible_tower_fqdn').find_each do |parameter|
+      parameter.update(
+        name: 'ansible_tower_api_url',
+        value: "https://#{parameter.value}/api/controller/v2"
```
I think this would make the param value default to AAP 2.5, but what If a user wishes to continue to using either AAP 2.3 or AAP 2.4 post-upgrade, I think this may not work for that use case, right?

nofaralfasi replied:
Correct. In that case the user would have to manually update the variable's value. 
It's documented in the documentation PR: https://github.com/theforeman/foreman-documentation/pull/3720/
